octobot_trading/exchanges/connectors/ccxt/ccxt_client_util.py

```diff
@@ -0,0 +1,149 @@
+    """
+    Exchange instance creation
+    :return: the created ccxt (pro, async or sync) client
+    """
+    client = None
+    is_authenticated = False
+    if not exchange_manager.exchange_only:
+        # avoid logging version on temporary exchange_only exchanges
+        logger.info(f"Creating {exchange_class.__name__} exchange with ccxt in version {ccxt.__version__}")
+    if exchange_manager.ignore_config or exchange_manager.check_config(exchange_name):
+        try:
+            key, secret, password = exchange_manager.get_exchange_credentials(logger, exchange_name)
+            if not (key and secret) and not exchange_manager.is_simulated:
+                logger.warning(f"No exchange API key set for {exchange_manager.exchange_name}. "
+                                    f"Enter your account details to enable real trading on this exchange.")
+            if should_authenticate():
+                client = exchange_class(_get_client_config(options, headers, additional_config, 
+                                                           key, secret, password))
+                is_authenticated = True
+                if exchange_manager.check_credentials:
+                    client.checkRequiredCredentials()
+            else:
+                client = exchange_class(_get_client_config(options, headers, additional_config))
+        except (ccxt.AuthenticationError, Exception) as e:
+            if unauthenticated_exchange_fallback is None:
+                return get_unauthenticated_exchange(
+                    exchange_class, options, headers, additional_config
+                ), False
+            return unauthenticated_exchange_fallback(e), False
+    else:
+        client = get_unauthenticated_exchange(exchange_class, options, headers, additional_config)
+        logger.error("configuration issue: missing login information !")
+    client.logger.setLevel(logging.INFO)
+    _use_http_proxy_if_necessary(client)
+    return client, is_authenticated
+
+
+def get_unauthenticated_exchange(exchange_class, options, headers, additional_config):
+    client = exchange_class(options, headers, additional_config)
+    _use_http_proxy_if_necessary(client)
+    return client
+
+
+def set_sandbox_mode(client, is_sandboxed):
+    client.setSandboxMode(is_sandboxed)
+
+
+def get_ccxt_client_login_options(exchange_manager):
+    """
+    :return: ccxt client login option dict, can be overwritten to custom exchange login
+    """
+    if exchange_manager.is_future:
+        return {'defaultType': 'future'}
+    if exchange_manager.is_margin:
+        return {'defaultType': 'margin'}
+    return {'defaultType': 'spot'}
+
+
+def get_symbols(client):
+    return set(client.symbols) \
+        if hasattr(client, "symbols") and client.symbols is not None else set()
+
+
+def get_time_frames(client):
+    return set(client.timeframes) \
+        if hasattr(client, "timeframes") and client.timeframes is not None else set()
```

Comment:
Should we use a try catch attribute error instead of checking with hasattr ?

Reply:
we would still need a if to check if it's None, I think we should leave it line this

Reply:
I don't understand why checking if value is None should be taken into account here.
I don't say using a try catch will simplify the code. My point is I think we should use hasattr only when it's necessary